proposals/003-authentication-api.md

```diff
@@ -0,0 +1,223 @@
+
+# Authentication API
+
+This proposal describes a public API to expose client identity to `Filter` implementations.
+
+## Current situation
+
+The `Filter` and `FilterContext` APIs currently don't directly expose any authenticated client identity information.
+Specifically:
+
+* If clients are authenticating to the proxy using mTLS then filters don't have access to a `Subject` or `Principal` corresponding to the client TLS certificate.
+* If clients are authenticating using the `SaslAuthenticate` KRPC the only way a `Filter` can know about that is by intercepting those frames.
+    - For "SASL passthrough", a `Filter` implementation could try to _infer_ the client's identity by watching for a successful `SaslAuthenticateResponse` 
+      returned by the broker. This is a technique that each identity-dependent filter in the chain could use, but each needs to reimplement the logic.
+    - For "SASL termination", a `Filter` implementation could handle the `SaslAuthenticateRequest` frames itself 
+        - If it makes its own `SaslAuthenticateRequests` (with different credentials, and/or a different mechanism) to the server then subsequent filters in the chain need to use the "SASL passthrough" technique to be aware of the broker-facing identity, and there's no way for them to know the original identity.
+        - If it doesn't (so that there is no `SaslAuthenticate` intersection with the broker) then subsequent filters in the chain have no way of knowing the client's identity.
+
+## Motivation
+
+The lack of this API makes implementing client identity aware plugins difficult, or impossible.
+
+Goals: 
+
+* Enable `Filters` to access a client's identity using a single, consistent API, irrespective of what authentication mechanism is implemented, TLS or SASL, and whether it's implemented by the proxy runtime (in the TLS case), or a prior Filter in the chain (in the SASL termination case).
+* Don't require a `Filter` to handle `SaslAuthenticate` unless it is performing SASL termination. 
+
+Non-goals:
+
+* Defining an API for exposing the identity of a _broker_ to `Filters` (in cases where the proxy mututally authenticates the broker).
+
+## Proposal
+
+### Public API changes
+
+Add the following interface to the `kroxylicious-api` module to allow a `Filter` implementation to opt into consuming authentication outcomes:
+
+```java
+package io.kroxylicious.proxy.filter;
+
+import javax.security.auth.Subject;
+import javax.security.auth.login.LoginException;
+
+// This is a new interface in the api module
+interface ClientSubjectAwareFilter extends Filter {
+    // Called when a client authenticates, or reauthenticates
+    void onClientAuthentication(Subject clientSubject, FilterContext context);
+    // Called when a client fails authentication, or reauthentication
+    default void onClientAuthenticationFailure(LoginException exception, FilterContext context) { 
+    }
```

Comment:
Is it a implementations job when overriding this to invoke `FilterContext.clientAuthenticationFailure`? or are you expecting the runtime to just send the same failure to each filter in the chain?

I'm wondering how we could/would support a filter noticing that a `onClientAuthenticationFailure` had happened and generating a new temporary token. Can it then make a out of band request to authenticate with that new token? When that completes does it call `clientAuthenticationSuccess`? Does it make sense in this scenario for the proxy to have told every other filter in the chain about the same failed authentication event?

Reply:
Closing as outdated since the conversation about principals/subjects is ongoing in fresher threads and the event based flow has been replaced with FilterContext methods